src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
-                                           final double HistogramMADTolerance) {
-
-        final JavaRDD<Integer> lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
-        lengthRDD.cache();
-
-        metrics.READ_PAIRS = filteredReads.count();
-
-        collectLevelInfo();
+                                           final SAMFileHeader header,
+                                           final Set<MetricAccumulationLevel> accumLevels,
+                                           final double histogramMADTolerance) {
+
+        // construct untrimmed hand rolled "histogram" (SortedMap) in three steps, because htsjdk Histogram does not play well with Serialization
+        // so first hand roll a histogram using Spark (read traversal is the bottleneck in terms of performance),
+        //   when computing corresponding metrics locally, use htsjdk Histogram converted the sorted map version to do the work
+        final Map<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> unsortedHistogramsAtRGLevel = filteredReads.mapToPair(read -> step1ReadTraversal(read, header))
+                                                                                                                                    .groupByKey()
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step2ConvertToList)
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step3ConvertToHistogram)
+                                                                                                                                    .collectAsMap();
+
+        final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtRGLevel = new TreeMap<>(groupInfoComparator);
+        histogramsAtRGLevel.putAll(unsortedHistogramsAtRGLevel);
+
+        LinkedList<SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>>> listOfStats = new LinkedList<>();
+
+        if (accumLevels.contains(MetricAccumulationLevel.READ_GROUP)) {
+            listOfStats.addFirst(histogramsAtRGLevel);
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.LIBRARY)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.LIBRARY));
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.SAMPLE)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.SAMPLE));
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.ALL_READS)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.ALL_READS));
+        }
 
-        collectMedianAndMAD(lengthRDD);
+        int i=0;
```

Comment:
You could change this to a for loop over i to listOfStats.keySet.size(), then you don't have to increment i as a side effect in the last line of this loop.